Stories: Refine filter feature #92

mynameisted opened the pull request:
Issue #92 has been split into 4 smaller sub-issues.

In this PR:
- Resolves #100 **Stories: Multiple selection for filtering** and resolves #98 **Stories: Display selected tags for filter**  in 98ac00275061eb05e3986359f2eda5ba94a6e468 by replacing the existing select drop-down with [bootstrap-select](https://github.com/snapappointments/bootstrap-select).
- Resolves #97 **Stories: Update sizing of filter button** in f65332c89fa93b2c65c1e8ce1bae7d20fba0b457 by increasing the font size of the select button.

Closing #92 as 3 of the sub-isuses has been resolved in this PR, the last sub-issue will be tracked separately.

This PR will invalidate the existing code for hiding certain post cards, thus the existing jQuery logic has been removed in f78bf64c551b0c0b19cd884af487856844f1e047.
This will be re-implemented in a future PR under #99  **Stories: Filtering not working as expected**.

wei2912 reviewed (commented):
Thanks for the PR, I'll take a closer look when I have time to test it out.

In the meantime, could the comments on `default.hbs` at L11 and L83 be deleted?

mynameisted commented:
> 
> 
> Thanks for the PR, I'll take a closer look when I have time to test it out.
> 
> In the meantime, could the comments on `default.hbs` at L11 and L83 be deleted?

Comments have been removed in 7982812.

wei2912 commented:
Thanks, I'll test out the filtering of tags again (maybe by tomorrow?) and let you know how it goes.

wei2912 commented:
Hi @mynameisted, I've just fixed the bug with the Stories page filtering, see https://github.com/AdvisorySG/ghost-advisory-theme/pull/106 for more details. You should be able to reimplement the rest of the filtering now.

wei2912 reviewed (changes requested):
To add in filter features after addressing https://github.com/AdvisorySG/ghost-advisory-theme/issues/99.

wei2912 commented:
@mynameisted Hello, I've just taken another look at your PR. Once you add back the filtering logic, let me know and I'll do a final code review before merging the PR. Thanks!

mynameisted commented:
@wei2912 Hi, I've added the custom multi-select dynamic filtering of posts with persisted filtering across paginated pages via localStorage.

wei2912 commented on default.hbs:
```diff
@@ -76,23 +73,123 @@
+                    let internalTag = $(this).hasClass('internal-tag');
+                    // Basic slug generator: replace whitespaces with dashes, then strip punctuation
+                    // TODO: Document behaviour of this generator in API v3
+                    classToFilter = classToFilter.replace(/\s+/g, '-').replace(/[.,'!?]/, '').toLowerCase();
```
Could this be abstracted into a separate function, since it was used on L104 and L178 (here)? To clarify my comment on L177, the conversion into a slug is not well-documented in the Ghost docs. So, if we reuse this regex it should be placed into a separate function with a note that this may break at any time.

mynameisted replied:
Fixed as part of f1e4169600a1c7305e809dc571697d11841dec8a

wei2912 commented on default.hbs:
```diff
@@ -76,23 +73,123 @@
+                    // Write values for internal tags
+                    internalFilters = filterOptions;
+                    // Write internal filters to localstorage for persistance across pages
+                    localStorage.setItem('internalStories', rawFilter.join("|"));
```
Could `internalStories` be changed to `internalFilters`? Since it's the tags of the stories, and not the stories themselves, which have an internal/primary distinction.

mynameisted replied:
Fixed as part of 62bf77a2732fba5747e5434c4110844a120c146d

wei2912 commented on default.hbs:
```diff
@@ -76,23 +73,123 @@
+                    // Write values for primary tags
+                    primaryFilters = filterOptions;
+                    // Write filters to localstorage for persistance across pages
+                    localStorage.setItem('primaryStories', rawFilter.join("|"));
```
See L110.

mynameisted replied:
Fixed as part of 62bf77a2732fba5747e5434c4110844a120c146d

wei2912 commented on default.hbs:
```diff
@@ -76,23 +73,123 @@
+                    classToFilter = classToFilter.replace(/\s+/g, '-').replace(/[.,'!?]/, '').toLowerCase();
+                    console.log("cleaned Filtering: " + classToFilter);
+
+                    $('.gh-postfeed').children().each(function() {
```
For L181-192, some of this code seems to be repeated/refined in the new functions you created (such as `filterPosts`). Would it be possible to simplify this part of the code to use these new functions?

mynameisted replied:
Fixed as part of d30d10e21d7a253652309c15692af5631695ac6a

wei2912 commented on default.hbs:
```diff
@@ -7,6 +7,8 @@
         <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/css/bootstrap.min.css" integrity="sha384-Vkoo8x4CGsO3+Hhxv8T/Q5PaXtkKtu6ug5TOeNV6gBiFeWPGFN9MuhOf23Q9Ifjh" crossorigin="anonymous">
         <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
+        <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-select@1.13.14/dist/css/bootstrap-select.min.css">
+        <link rel="stylesheet" type="text/css" href="{{asset "built/screen.css"}}" />
```
I think `built/main.css` should already include `screen.css` as part of its build process.

mynameisted replied:
Fixed as part of bf644fb85f0e1440e7dd05501937f1827868059c

wei2912 commented on partials/pagination.hbs:
```diff
@@ -0,0 +1,20 @@
+<script type="text/javascript">
```
Could this file be modified to use 4-space indentation?

mynameisted replied:
Fixed as part of be822a9f59639f12f436a55cedc415794b85a0d2

wei2912 reviewed (changes requested):
Thanks for the code! The logic looks good, but unfortunately I wasn't able to test it due to this error:

![image](https://user-images.githubusercontent.com/2070423/107855169-c28c8380-6e5b-11eb-9f17-76141048aa55.png)

Do you have any idea why this is the case?

mynameisted commented:
> 
> 
> Thanks for the code! The logic looks good, but unfortunately I wasn't able to test it due to this error:
> 
> ![image](https://user-images.githubusercontent.com/2070423/107855169-c28c8380-6e5b-11eb-9f17-76141048aa55.png)
> 
> Do you have any idea why this is the case?

I believe you have to update your routes.yaml file to support pagination on the stories page. I have copied the contents of my routes.yaml file below for your reference.

**Routes.yaml**
```
routes:

collections:
  /:
    permalink: /{slug}/
    template: index
  /stories/:
    permalink: /{slug}/
    template: page-stories

taxonomies:
  tag: /tag/{slug}/
  author: /author/{slug}/
```

wei2912 commented:
> I believe you have to update your routes.yaml file to support pagination on the stories page. I have copied the contents of my routes.yaml file below for your reference.

Thanks! I wasn't aware of this...

I was testing the pagination + filter and found a few issues (for reference, the setup I am using has the exact same posts as beta.advisory.sg):

1. On my setup, the pagination is split into 3 pages (each page with 2 posts), although there are only 4 posts in total:

![image](https://user-images.githubusercontent.com/2070423/108840667-aafa8b00-7611-11eb-89df-15360ce89949.png)

2. As the pagination uses Ghost's native pagination feature, it is possible to set the filter such that no posts appear on a page:

![image](https://user-images.githubusercontent.com/2070423/108840904-f01ebd00-7611-11eb-9a28-7c15badeb392.png)

Ideally, the pagination should be changed such that each page has the same number of posts (other than the last page). I don't think this is possible with Ghost's native pagination, so we may have to seek other ways e.g. loading content from Ghost API and dynamically allocating to pages.

3. Deselecting all the options is equivalent to selecting all the options. I can see why it's implemented in this manner, but I think it's best to stick to deselecting all = display no results, selecting all = display all results. This way would be more intuitive for users.

bnjmnt4n commented:
> I believe you have to update your routes.yaml file to support pagination on the stories page. I have copied the contents of my routes.yaml file below for your reference.

Maybe it's a good idea to upload a template for the `routes.yaml` (and maybe any other configuration files needed for a properly functioning dev environment) somewhere?

mynameisted commented:
> 
> 
> > I believe you have to update your routes.yaml file to support pagination on the stories page. I have copied the contents of my routes.yaml file below for your reference.
> 
> Thanks! I wasn't aware of this...
> 
> I was testing the pagination + filter and found a few issues (for reference, the setup I am using has the exact same posts as beta.advisory.sg):
> 
>     1. On my setup, the pagination is split into 3 pages (each page with 2 posts), although there are only 4 posts in total:
> 
> 
> ![image](https://user-images.githubusercontent.com/2070423/108840667-aafa8b00-7611-11eb-89df-15360ce89949.png)
> 
>     1. As the pagination uses Ghost's native pagination feature, it is possible to set the filter such that no posts appear on a page:
> 
> 
> ![image](https://user-images.githubusercontent.com/2070423/108840904-f01ebd00-7611-11eb-9a28-7c15badeb392.png)
> 
> Ideally, the pagination should be changed such that each page has the same number of posts (other than the last page). I don't think this is possible with Ghost's native pagination, so we may have to seek other ways e.g. loading content from Ghost API and dynamically allocating to pages.
> 
>     1. Deselecting all the options is equivalent to selecting all the options. I can see why it's implemented in this manner, but I think it's best to stick to deselecting all = display no results, selecting all = display all results. This way would be more intuitive for users.

Hi @wei2912 , I have fixed the following issues in your review:
1. Fixed pagination numbering error, it should be working now.
2. Agreed, this was raised up to you previously when contemplating and deciding on reusing the 'dynamic filtering' approach for filtering the stories. Ghost supports content hubs called Channels, which provides filtered and paginated content of a specific tag, which may be a potential area to be looked into if so.
3. I've updated the filtering logic to hide all posts if no options are selected. If the user visits the page for the **first time and the localStorage variables are not set** yet, all options in the filter are selected by default so that the user does not land on an empty stories page on first load.

mynameisted commented:
> 
> 
> > I believe you have to update your routes.yaml file to support pagination on the stories page. I have copied the contents of my routes.yaml file below for your reference.
> 
> Maybe it's a good idea to upload a template for the `routes.yaml` (and maybe any other configuration files needed for a properly functioning dev environment) somewhere?

Hi @bnjmnt4n , agreed, but as this repo consists of only the custom ghost theme, it is not advisable to include the routes.yaml fle in this repo due to the file structure of the ghost platform. Perhaps the routes.yaml structure/file can be sent into the telegram group chat instead for all developers to update on their respective ends?

wei2912 commented:
> Hi @bnjmnt4n , agreed, but as this repo consists of only the custom ghost theme, it is not advisable to include the routes.yaml fle in this repo due to the file structure of the ghost platform. Perhaps the routes.yaml structure/file can be sent into the telegram group chat instead for all developers to update on their respective ends?

On `routes.yaml`, the file is downloadable on beta.advisory.sg/ghost so I don't think it'll be a big issue. In the future I hope to provide an easier way of downloading the site's configurations & content (will need to work with @AdvisorySG/sysadmins).

> Agreed, this was raised up to you previously when contemplating and deciding on reusing the 'dynamic filtering' approach for filtering the stories. Ghost supports content hubs called Channels, which provides filtered and paginated content of a specific tag, which may be a potential area to be looked into if so.

An issue with Channels is that they don't allow people to filter for both industry and article type, and select multiple options (unless we decide to do all possible combinations, which will be very painful). I think it's likely that users will need to select multiple industry types, when sorting through articles on relevant industries. (I'm not sure if users would do this for article type, but I guess it can't hurt to provide that functionality.)

In any case, I think this PR ought to be merged first while we explore alternative approaches. I just have one last comment - the logic on https://github.com/AdvisorySG/ghost-advisory-theme/pull/102/commits/ab19663cb57680c4e67861c01824f7fcccdb90ea#diff-c38171c984efe4c24f5b7920d3e9bb8c53d7819a87c4527d397b018448965e0cR129 means that posts will be shown if they are selected by either of the filters. Could we change this behaviour to posts meeting the criteria of both filters? For example, if one selects no industries and all article types, they will receive the full list of articles:

![Uploading image.png…]()

mynameisted commented:
> 
> 
> > Hi @bnjmnt4n , agreed, but as this repo consists of only the custom ghost theme, it is not advisable to include the routes.yaml fle in this repo due to the file structure of the ghost platform. Perhaps the routes.yaml structure/file can be sent into the telegram group chat instead for all developers to update on their respective ends?
> 
> On `routes.yaml`, the file is downloadable on beta.advisory.sg/ghost so I don't think it'll be a big issue. In the future I hope to provide an easier way of downloading the site's configurations & content (will need to work with @AdvisorySG/sysadmins).
> 
> > Agreed, this was raised up to you previously when contemplating and deciding on reusing the 'dynamic filtering' approach for filtering the stories. Ghost supports content hubs called Channels, which provides filtered and paginated content of a specific tag, which may be a potential area to be looked into if so.
> 
> An issue with Channels is that they don't allow people to filter for both industry and article type, and select multiple options (unless we decide to do all possible combinations, which will be very painful). I think it's likely that users will need to select multiple industry types, when sorting through articles on relevant industries. (I'm not sure if users would do this for article type, but I guess it can't hurt to provide that functionality.)
> 
> In any case, I think this PR ought to be merged first while we explore alternative approaches. I just have one last comment - the logic on [ab19663#diff-c38171c984efe4c24f5b7920d3e9bb8c53d7819a87c4527d397b018448965e0cR129](https://github.com/AdvisorySG/ghost-advisory-theme/commit/ab19663cb57680c4e67861c01824f7fcccdb90ea#diff-c38171c984efe4c24f5b7920d3e9bb8c53d7819a87c4527d397b018448965e0cR129) means that posts will be shown if they are selected by either of the filters. Could we change this behaviour to posts meeting the criteria of both filters? For example, if one selects no industries and all article types, they will receive the full list of articles:
> 
> ![image](https://user-images.githubusercontent.com/2070423/108871199-ddb57b00-7633-11eb-94e1-4b6797f154d0.png)

@wei2912  As requested, I've amended the filter behavior to cater for AND relationships if there are both primary and internal filters selected as part of 1393fac37d9d5efe30be86c39801d2ec39a4575d.

mynameisted commented:
Hi @wei2912 , as discussed, PR updated with a new pagination partial which should resolve the previous visual bugs. 


![image](https://user-images.githubusercontent.com/42962723/112289981-c10b7200-8cc9-11eb-858c-dd15ec4da204.png)
![image](https://user-images.githubusercontent.com/42962723/112290159-ea2c0280-8cc9-11eb-8c40-f5c3314aa991.png)

wei2912 commented:
@mynameisted I've merged this PR in https://github.com/AdvisorySG/ghost-advisory-theme/commit/b1e83d606a6cef233f6eed3286cb3e075c530173, thanks a lot! After this, we can start working on improving the pagination by loading the posts dynamically from the Ghost Content API.
